### ISSUE TITLE
[BUG] Inconsistent behaviour of Ansible modules (update vs remove + compared to API)

### ISSUE THREAD

**Issue by Dennis-Q** [Checkmk/ansible-collection-checkmk.general#533]
**Describe the bug**
The behaviour of the ansible module 'folder', and possibly other modules too which can apply labels, seem to have an inconsistent behavior. There are actually a few different 'issues' with it. Let me try to explain the differences and what we would expect from them.

**Inconsistency between update_attribute and remove_attribute within ansible module**
1. remove_attributes is not removing specific labels (and just states 'OK' instead)
2. remove_attributes is not giving an error when a dictionary is provided (it seems to only work with a list)

**Inconsistency between update_attributes of checkmk module vs API-behavior (update_attributes)**
1. When you use update_attributes to add a label using the Ansible module, it actually adds the label **in addition to** the existing labels. The API however, replaces the complete set of labels. This has been communicated with multiple employees of the Checkmk company.
2. remove_attributes seems to be in-line with the API's behavior; it was reported back to us that it works as intended and there were no plans to only remove specific labels (unless a feature request would be created).

**Component Name**

Component Name: folder_module

**Ansible Version**

```console
$   ansible --version
ansible [core 2.14.11]
  config file = /home/user/repos/ansible-checkmk/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/venvs/ansible-checkmk/lib64/python3.9/site-packages/ansible
  ansible collection location = /home/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/user/venvs/ansible-checkmk/bin/ansible
  python version = 3.9.16 (main, May 29 2023, 00:00:00) [GCC 11.3.1 20221121 (Red Hat 11.3.1-4)] (/home/user/venvs/ansible-checkmk/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

**Checkmk Version and Edition**
```console
Checkmk Enterprise Edition 2.2.0p17
```

**Collection Version**

```console
$ ansible-galaxy collection list
checkmk.general        4.0.0 
```

**To Reproduce**
1. Manually create a folder and add a label. e.g.: 'pretest:label'
2. Create task/play to use ansible module 'folder' using the 'update_attributes' to add a label.  example:
```yaml
---
- name: "Create/update folder"
  checkmk.general.folder:
    server_url: "{{ master_site_baseurl }}"
    site: "{{ master_site }}"
    automation_user: "{{ automation_user }}"
    automation_secret: "{{ automation_secret }}"
    path: "/{{ item.path.lower() }}"
    name: "{{ item.name }}"
    update_attributes:
      labels: "{{ item.labels|default(omit) }}"
    state: "{{ item.state | default('present') }}"
  delegate_to: localhost
  loop: 
    - { name: "Testdir", path: "testdir", labels: { "custom/device_type": "network" } }
```
3. Run the play.
4. Check result; it now contains 2 labels: 'pretest:label' and 'custom/device_type:network', which is 'as expected'.
5. Add new play to remove label, similar to the above (by replacing 'update_attributes' with 'remove_attributes'), example:
```yaml
- name: "Remove label"
  checkmk.general.folder:
    server_url: "{{ master_site_baseurl }}"
    site: "{{ master_site }}"
    automation_user: "{{ automation_user }}"
    automation_secret: "{{ automation_secret }}"
    path: "/{{ item.path.lower() }}"
    name: "{{ item.name }}"
    remove_attributes:
      labels: "{{ item.labels|default(omit) }}"
    state: "{{ item.state | default('present') }}"
  delegate_to: localhost
  loop: 
    - { name: "Testdir", path: "testdir", labels: { "custom/device_type": "network" } }
```
6. Run the play; in this case ansible just does nothing and states 'OK'. We would either expect it to remove the specified label (similar to how update_label worked) or give an error (not accepting a dict) or another reason why it didn't do anything... **Stating 'OK' in this case is wrong.** 
7. When you update the task to include 'labels' as a list instead (not providing the actual labels through a dictionary), you can see it removes all the labels (not specific ones only) from the folder. Example:
```yaml
- name: "Remove all labels"
  checkmk.general.folder:
    server_url: "{{ master_site_baseurl }}"
    site: "{{ master_site }}"
    automation_user: "{{ automation_user }}"
    automation_secret: "{{ automation_secret }}"
    path: "/{{ item.path.lower() }}"
    name: "{{ item.name }}"
    remove_attributes:
    - labels
    state: "{{ item.state | default('present') }}"
  delegate_to: localhost
  loop: 
    - { name: "Testdir", path: "testdir", labels: { "custom/device_type": "network" } }
```
8. Run the play and find out this removes all the labels (and checkbox) from the folder.
9. Do some testing with the API; you will see that update_attributes of the API will just replace all labels with the newly provided ones. It does NOT retain any existing labels like the ansible module does. I think this is still being discussed internally within the Checkmk company. We would only expect this to happen when we use the 'attributes' option instead of 'update_attributes'.

**Expected behavior**
The current behaviour of the Ansible module we see when using update_attributes is the most feasible and flexible; it allows you to add labels and not replacing them or first having to find out which ones are already present. The same behaviour was expected with the remove_attributes. However, in practice nothing happens when specifying labels using a dict. In our opinion it should just have removed that specific label (and retain any other already set to the folder).

Also, this behavior is not in-line with the API which can be confusing. Still we prefer the way the ansible module works.

**Actual behavior**
remove_attributes when providing a dict with labels states 'OK' and doesn't execute anything. When providing 'labels' as a list (without any actual labels) it does remove all the labels from the folder. This last part is in-line with the API behavior, not necessarily what we would expect to happen. 

**Minimum reproduction example**
See above

**Comment by msekania** [comment]
@Dennis-Q,

I initially misinterpreted the question.

Let me do some brainstorming.

If I have understood correctly. Let's say

1. if one provides a list, then the module should behave similarly to the API and remove top level attributes with the keys specified in the list. (the module behaves as one would expect from the API),
2. if one provides a dict instead then it should take the difference between the current attributes and the dict provided. (definitely tricky to implement)

For 2, just to get the logic right, let's assume that we have
```yaml
abc:
  "A": 
    "B":
      "C"
    "D":
      "E":
        "F"
    "G":
       "H":
         "I"
  "D":
    "E":
      "F"

cba:
  "A": 
    "B":
      "C"
    "D":
      "E":
        "F"
    "G":
       "H":
         "X"
    "1":
       "2"
         "3"
```

then for `curr = abc` and `par = cba` one should remove:
```yaml
  "A":
    "G":
      "H":
        "I"
  "D":
    "E":
       "F"
```

and for `curr = cba` and `par = abc` one should remove:
```yaml
  "A":
    "1":
      "2":
        "3"
    "G":
      "H":
        "X"
```
if that's a desired behavior I think it should be doable, at least on python 3.10 level, for "Ansible" not sure yet. If subset of Python in Ansible will not cause the problem, I can implement this logic.

**Comment by Dennis-Q** [comment]
Hi @msekania ,

`1. if one provides a list, then the module should behave similarly to the API and remove top level attributes with the keys specified in the list. (the module behaves as one would expect from the API),`
This is true for remove_attributes (seems identical to API), but not for update_attributes (Ansible adds to existing labels, API replaces all with the provided list)
`2. if one provides a dict instead then it should take the difference between the current attributes and the dict provided. (definitely tricky to implement)`
Remark: the API doesn't allow you to provide a dictionary for the remove_attributes. The ansible module seems to 'accept' it (no errors) but doesn't actually handle them. I indeed would expect that the items (labels) provided in the remove_attributes list are removed from the existing labels already present, so I think the answer to your question is yes here.

I am not entirely sure about your explanation, but maybe that's due to my lack of knowledge of Python/terminology. Maybe we can create an example with actual labels? I only know the attribute names and ansible/jinja2 terminology for things. I asked my colleague but he was also a bit confused/unsure at first glance.


Also, I think there are 3 issues mentioned here:

- A. Ansible remove_attributes not giving error when providing dict (but also not executing anything)

- B. Ansible remove_attributes working different from update_attributes (update_attributes add labels to existing labels, remove_attributes only seem to be able to remove all labels (and not specific ones)).

- C. Different behaviour between Ansible and API (e.g. API will never keep existing labels, but just replaces it with the newly defined set when using update_attributes. remove_attributes when using a list seems to work the same)

Maybe it's good to be more specific about which issue we're talking about.


One more note: I think this issue is specific for labels. These differentiate from other attributes as you can provide multiple instead of one (predefined) value.

I hope this makes a little more sense now. Thank you for the quick follow-up.

**Comment by msekania** [comment]
Hi @Dennis-Q,

I was already talking how one can extend the extend the existing functionality.

I know how it works currently, and yes it might be not what user expects directly, but is not certainly a bug, rather inconsistent behavior.

Anyway, my 1. an 2. points were how I understood your "Expected behavior" (your proposal).

**Comment by Dennis-Q** [comment]
> Hi @Dennis-Q,
> 
> I was already talking how one can extend the extend the existing functionality.
> 
> I know how it works currently, and yes it might be not what user expects directly, but is not certainly a bug, rather inconsistent behavior.
> 
> Anyway, my 1. an 2. points were how I understood your "Expected behavior" (your proposal).

So I fully agree with these points if it is about the remove_attributes option.

Of course, you could decide to stick with the API behavior or decide to adapt the update_attribute option to make it in line with the API.

Happy you are willing to adapt to what we suggested! It was suggested by @lgetwan to make this issue as we debugger this behaviour together with him.

Would be happy to test when a new version is available!

**Comment by msekania** [comment]
@Dennis-Q,

just to clarify things,  from e.g., 2.2.0p16
![image](https://github.com/Checkmk/ansible-collection-checkmk.general/assets/113431380/c1acaa79-541c-4c8d-9775-d79b09363989)

In essence, currently  update_attributes does what API claims and also remove_attributes if list is provided as an input (dict is just converted to list). 

Anyway, I am preparing extention

**Comment by Dennis-Q** [comment]
> @Dennis-Q,
> 
> just to clarify things, from e.g., 2.2.0p16 ![image](https://private-user-images.githubusercontent.com/113431380/300195245-c1acaa79-541c-4c8d-9775-d79b09363989.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYzNzcxODcsIm5iZiI6MTcwNjM3Njg4NywicGF0aCI6Ii8xMTM0MzEzODAvMzAwMTk1MjQ1LWMxYWNhYTc5LTU0MWMtNGM4ZC05Nzc1LWQ3OWIwOTM2Mzk4OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyN1QxNzM0NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hM2VjMjZlNWY4NWE1NzNkYWFlYWVkZmMzMjFiMGYyYzJkMGNhMDVhZjczM2RmNjExZjQ5YzExZTYwMjRiNDQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4Dm4OAyiyWKrmQTVejS64eYGWCuoRPcVrSY7Wz5NUXA)
> 
> In essence, currently update_attributes does what API claims and also remove_attributes if list is provided as an input (dict is just converted to list).
> 
> Anyway, I am preparing extention

Currently, the module does not behave the same way for labels as the latest checkmk (including the one you mentioned).

Please create a folder with a label, and run the ansible module with update_attributes to change/add labels. Then do the same thing using the API instead.

I am pretty sure you understand what happens when you actually try this as per my description.

**Comment by msekania** [comment]
Hi @Dennis-Q, 

check #539

**Comment by Dennis-Q** [comment]
@msekania Thank you for sharing. The code changes look good at first sight; I am happy to test the end-result when it is commited to a branch and see if it works as expected.

**Comment by robin-checkmk** [comment]
Just to summarize: The changes proposed by @msekania in #539 would solve this issue, correct?
So I would go ahead and link the PR and this issue, so we know when to close it.
